File: epi_judge_python/swap_bits.py
from test_framework import generic_test


def swap_bits(x, i, j):
    # TODO - you fill in here.
    # ib = (x & 1 << i == 0)
    # jb = (x & 1 << j == 0)
    # if ib == jb:
    #     return x
    # if ib:
    #     x |= 1 << i 
    # else:
    #     x &= ~(1 << i)
    # if jb:
    #     x |= 1 << j 
    # else:
    #     x &= ~(1 << j)
    # return x
    if (x >> i & 1) != (x >> j & 1):
        x ^= 1 << i | 1 << j
    return x

if __name__ == '__main__':
    exit(
        generic_test.generic_test_main('swap_bits.py', 'swap_bits.tsv',
                                       swap_bits))

File: epi_judge_cpp/swap_bits.cc
#include "test_framework/generic_test.h"
long long SwapBits(long long x, int i, int j) {
  if (i == j) {
    return x;
  }
  long long one = 1;
  short ib = (x >> i) & one;
  short jb = (x >> j) & one;
  // std::cout << std::bitset<64>(x) << std::endl;
  if (ib != jb) {
    x ^= one << i;
    x ^= one << j;
  }
  return x;
}

int main(int argc, char *argv[]) {
  std::vector<std::string> args{argv + 1, argv + argc};
  std::vector<std::string> param_names{"x", "i", "j"};
  return GenericTestMain(args, "swap_bits.cc", "swap_bits.tsv", &SwapBits,
                         DefaultComparator{}, param_names);
}
